### PR TITLE
[mtouch] Add `force-rejected-types-removal` optimization

### PR DESCRIPTION
**Pull request by spouliot**
This optimization can be enabled when it's not possible to use the
managed linker (e.g. **Don't link**) or when the managed linker cannot
remove references to deprecated types that would cause an application
to be rejected by Apple.

References to the existing types will be renamed, e.g. `UIWebView` to
`DeprecatedWebView`, in every assemblies.

The type definition is also renamed (for validity) and all custom
attributes on the types and their members will be removed.
Code inside the members will be replaced with a
`throw new NotSupportedException ();`.

The msbuild test app `MyReleaseBuild` has been updated to test that the
optimization is working as expected (device builds are slow so reusing
this test has little impact in test time).

Basically the test ensure that `UIWebView` is used and cannot be removed
by the compiler (optimization) or the managed linker (since it's
referenced). Since the optimization is enabled then we can `grep` then
final `.app` directory to ensure there's no mention of `UIWebView` inside
any of the files that would be submitted.

The application can be run, by itself, and will turn green if OK, red if
`DeprecatedWebView` can't be found (skeleton replacement for `UIWebView`)
or orange if a `NotSupportedException` is thrown.

Finally introspection tests have been updated to skip over the deprecated
(and renamed) types. It should not be an issue right now, since this
optimization is not enabled by default, but it made testing easier.

### CODE REVIEW

File: docs/website/optimizations.md [hunk 1]
```diff
@@ -769,3 +769,21 @@ a product assembly and does not automagically mark all `NSObject`
...
+References to the existing types will be renamed, e.g. `UIWebView` to 
+`DeprecatedWebView`, in every assemblies.
+
+The type definition is also renamed (for validity) and all custom 
+attributes on the types and their members will be removed. 
+Code inside the members will be replaced with a
+`throw new NotSupportedException ();`.
+
+The default behavior can be overridden by passing
+`--optimize=[+|-]force-rejected-types-removal` to `mtouch`.
```
Comment (dalexsoto) [inline docs/website/optimizations.md]: We should put a list of the removed types here or a link to where it lives in our source code

File: tools/linker/MonoTouch.Tuner/RemoveCodeBase.cs [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+						static_ctor = method;
+					else
+						ProcessMethod (method);
+				}
+				if (static_ctor != null)
+					type.Methods.Remove (static_ctor);
+			}
+		}
+
+		new protected virtual void ProcessMethod (MethodDefinition method)
```
Comment (dalexsoto) [inline tools/linker/MonoTouch.Tuner/RemoveCodeBase.cs]: `new protected virtual` BOLD! 

/me hides

File: docs/website/optimizations.md (hunk 1 above)
Comment (spouliot) [reply docs/website/optimizations.md]: a link is a good idea - since it would be too easy to get out of sync

File: tools/linker/MonoTouch.Tuner/RemoveCodeBase.cs (hunk 2 above)
Comment (spouliot) [reply tools/linker/MonoTouch.Tuner/RemoveCodeBase.cs]: heh, just moving code (but it might be mine from old times)

File: tools/linker/RemoveRejectedTypesStep.cs [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+			base.Process (assembly);
+
+			var updated = false;
+			if (Profile.IsProductAssembly (assembly)) {
+				updated = true;
+				foreach (var remove in ReferencesToBeRemoved) {
+					var type = assembly.MainModule.GetType (remove);
+					if (type == null)
+						continue;
+					type.Name = type.Name.Replace ("UI", "Deprecated");
```
Comment (rolfbjarne) [inline tools/linker/RemoveRejectedTypesStep.cs]: This only works for types named `UI*`. Maybe instead remove the first two characters? Or maybe make `ReferencesToBeRemoved` a hash table where the lookup value is the string to search for?

The whole purpose of this code falls down if this replacement fails, so it might make sense to detect if something wasn't replaced as expected (for instance check that pre- and post-replacement-strings are not identical) and show a warning in that case.

File: tools/linker/RemoveRejectedTypesStep.cs [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+					Context.AddLinkedAwayType (type);
+					// remove [Register], [Protocol] and [ProtocolMembers] attributes
+					type.CustomAttributes.Clear ();
```
Comment (rolfbjarne) [inline tools/linker/RemoveRejectedTypesStep.cs]: You remove all attributes, but the comment sounds like you're only removing a few.

File: tools/linker/RemoveRejectedTypesStep.cs [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+					if (!type.IsInterface) {
+						// remove IL and custom attributes (e.g. [Export])
+						ProcessMethods (type);
+					}
+				}
+			} else if (Profile.IsSdkAssembly (assembly)) {
+				// note: we know the BCL does not include references to UIWebView
```
Comment (rolfbjarne) [inline tools/linker/RemoveRejectedTypesStep.cs]: This might not be correct when something new is added to the list of types to remove. Not sure what to do here though. One solution would be to (again) make `ReferencesToBeRemoved` a hash table with a boolean lookup value that states whether the type can be found in SDK assemblies or not.

File: tools/linker/RemoveRejectedTypesStep.cs [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+
+		protected override void ProcessMethod (MethodDefinition method)
+		{
+			if (method.HasCustomAttributes)
+				method.CustomAttributes.Clear ();
+
+            // don't break IDisposable
+//			if (method.Name != "Dispose") {
+				// deal with method `EnsureUIWebViewDelegate`
+				method.Name = method.Name.Replace ("UIWebView", "Deprecated");
```
Comment (rolfbjarne) [inline tools/linker/RemoveRejectedTypesStep.cs]: Indentation is wrong here.

File: tools/linker/RemoveRejectedTypesStep.cs [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+			// order is important for nested types
+			"UIKit.UIWebView/_UIWebViewDelegate",
+			"UIKit.UIWebView/UIWebViewAppearance",
+			"UIKit.IUIWebViewDelegate",
+			"UIKit.UIWebView",
+			"UIKit.UIWebViewDelegate",
+			"UIKit.UIWebViewDelegate_Extensions",
+			"UIKit.UIWebViewDelegateWrapper",
+			"UIKit.UIWebViewNavigationType",
+		};
```
Comment (rolfbjarne) [inline tools/linker/RemoveRejectedTypesStep.cs]: Would it make sense to make this list user-configurable?

Maybe something like: `--optimize=force-rejected-types-removal=type1,type2,...`

That might make it easier for us the next time Apple deprecates something (we can just tell customers the mtouch argument they have to use to get around the requirement until we get an update out with new entries in this list).

File: tools/linker/RemoveRejectedTypesStep.cs (hunk 4 above)
Comment (spouliot) [reply tools/linker/RemoveRejectedTypesStep.cs]: The list are the one that needs to be removed (at least so far)
I'll update the comment

File: tools/linker/RemoveRejectedTypesStep.cs (hunk 7 above)
Comment (spouliot) [reply tools/linker/RemoveRejectedTypesStep.cs]: That was by conscious design :) this is not a general purpose tool to remove code. There's way too many things that could go wrong (it was not as easy as expected, nor do I think it covers every possible cases) and the lure to abuse might be too strong ;-)

File: tools/linker/RemoveRejectedTypesStep.cs (hunk 3 above)
Comment (spouliot) [reply tools/linker/RemoveRejectedTypesStep.cs]: It was not meant as a general purpose mechanism (since it's not clear if the next stone to hit us would fit the same pattern). However it means one less place to update if we supply to alternative name.

File: tools/linker/RemoveRejectedTypesStep.cs (hunk 5 above)
Comment (spouliot) [reply tools/linker/RemoveRejectedTypesStep.cs]: This situation is very uncommon - it's the first time, after a dozen years, that a required removal affect our public API. All other times we only had to change our internals to avoid the _rejected_ API.

I don't think we need the extra complexity. Worse case we can revisit the code if/when that situation change but my guess is that, because of naming convention, the public BCL types are unlikely to be a trigger.

File: tools/linker/RemoveRejectedTypesStep.cs (hunk 3 above)
Comment (spouliot) [reply tools/linker/RemoveRejectedTypesStep.cs]: hmm... List<T> guaranteed order (so inner types could be processed before their declared types). That's not true for a hash table or a (standard) dictionary.

File: tools/linker/RemoveRejectedTypesStep.cs (hunk 6 above)
Comment (spouliot) [reply tools/linker/RemoveRejectedTypesStep.cs]: forgot that part - at some point _of code_ I had issue for `Dispose` :)

File: tools/linker/RemoveRejectedTypesStep.cs (hunk 3 above)
Comment (rolfbjarne) [reply tools/linker/RemoveRejectedTypesStep.cs]: You could also create a list of tuples instead.

Comment (spouliot) [reply tools/linker/RemoveRejectedTypesStep.cs]: yeah, that's what I'm doing. Time to move `mtouch` out of .net 4.6 :)
